cloudant-sync-datastore-core/src/test/java/com/cloudant/sync/replication/PushReplicatorTest.java

```diff
@@ -183,8 +185,135 @@ public void replicatorBuilderAddsCookieInterceptorDefaultPort() throws Exception
                         toString()
         );
         assertCookieInterceptorPresent(p, "name=%F0%9F%8D%B6&password=%F0%9F%8D%B6");
+    }
+
+    @Test
+    public void testPushReplicationFilter() throws Exception {
+        prepareTwoDocumentsInLocalDB();
+        ReplicatorBuilder.Push push = this.getPushBuilder();
+        push.filter(new PushFilter() {
+            @Override
+            public boolean shouldReplicateDocument(DocumentRevision revision) {
+                return revision.getBody().asMap().get("name").equals("Tom");
+            }
+        });
+        Replicator replicator = push.build();
+        replicator.start();
+        while (replicator.getState() != Replicator.State.COMPLETE && replicator.getState() !=
+                Replicator.State.ERROR) {
+            Thread.sleep(50);
+        }
+
+        Assert.assertEquals(Replicator.State.COMPLETE, replicator.getState());
+
+        // Check that the remote only contains the single doc we need.
+        Assert.assertEquals(1, couchClient.getDbInfo().getDocCount());
+    }
+
+    @Test
+    public void testPushReplicationFilterPushesZeroDocs() throws Exception {
+        prepareTwoDocumentsInLocalDB();
+        ReplicatorBuilder.Push push = this.getPushBuilder();
+        push.filter(new PushFilter() {
+            @Override
+            public boolean shouldReplicateDocument(DocumentRevision revision) {
+                return false;
+            }
+        });
+        Replicator replicator = push.build();
+        replicator.start();
+        while (replicator.getState() != Replicator.State.COMPLETE && replicator.getState() !=
+                Replicator.State.ERROR) {
+            Thread.sleep(50);
+        }
+
+        Assert.assertEquals(Replicator.State.COMPLETE, replicator.getState());
+
+        // Check that the remote contains no docs.
+        Assert.assertEquals(0, couchClient.getDbInfo().getDocCount());
+    }
+
+    /**
+     * This test checks that the replication continues from the correct place.
+     * It does this by first replicating with a filter that excludes all docs. This means that the 2
+     * local changes will get processed, but no documents will be added to the remote.
+     * Then it does another replication <b>without a filter</b>. We have made no more local changes
+     * so if the replicator starts from the correct checkpoint then there will be no changes and no
+     * more documents will be copied.
+     * The test will fail if the replicator starts from the wrong checkpoint because the 2 original
+     * changes would be visible after the second start() call.
+     *
+     * @throws Exception
+     */
+    @Test
+    public void testPushReplicationFilterContinuesFromCorrectPlace() throws Exception {
```

Comment:
Can you do asserts on the last saved sequence after each replication for these tests? Might be useful.